smfwatchdog.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SMFWATCHDOG" "1" "June 2013" "" "SMF Watchdog"
.
.SH "NAME"
\fBsmfwatchdog\fR \- SMF Health Checking Daemon
.
.P
A health checking daemon to be used with SMF services\.
.
.IP "\(bu" 4
\fISynopsis\fR
.
.IP "\(bu" 4
\fIInstallation\fR
.
.IP "\(bu" 4
\fIExample\fR
.
.IP "\(bu" 4
\fIOptions\fR
.
.IP "\(bu" 4
\fIUsage\fR
.
.IP "\(bu" 4
\fINotes\fR
.
.IP "\(bu" 4
\fITodo\fR
.
.IP "\(bu" 4
\fILicense (MIT)\fR
.
.IP "" 0
.
.P
 \fI\fR
.
.SH "Synopsis"
SMF manages services on Illumos \fIhttp://illumos\.org\fR based operating systems\. It ensures that services, which may consist of 0 to many processes, are up and running, and handles any failures they may encounter\.
.
.P
The one thing that SMF is lacking however, is the ability to do service level health checks\. SMF bases your services health on process health, checking things such as processes dying, or exiting abnormally, etc\. However, it may be the case that a process is up, has a pid and \fB/proc\fR structure, but has locked up, or is not behaving how it should\. \fBsmfwatchdog\fR is meant to do health checks (in the form of scripts) on your services, and proactively kill or restart the service under failure conditions\.
.
.P
The health check daemon is meant to run as a separate instance for each service you would like to monitor\. For example, an instance of \fBsmfwatchdog\fR would be running under the \fBnginx\fR service, while a different instance of \fBsmfwatchdog\fR would be running under the \fBapache\fR service\. The watchdog instance would only be responsible for checking the service under which it is running\.
.
.P
You can learn more about SMF by reading smf(5) \fIhttp://illumos\.org/man/5/smf\fR\.
.
.P
 \fI\fR
.
.SH "Installation"
Clone this repo, compile the source code, and install
.
.IP "" 4
.
.nf

git clone git://github\.com/bahamas10/smfwatchdog\.git
cd smfwatchdog
make
[sudo] make install
.
.fi
.
.IP "" 0
.
.P
This will install the \fBsmfwatchdog\fR binary and the \fBsmfwatchdog\.1\fR manpage\.
.
.P
 \fI\fR
.
.SH "Example"
A service that is a single Node\.JS \fIhttp://nodejs\.org\fR process acting as a webserver with the given SMF manifest is running on a machine\.
.
.P
\fBdaves\-service\.xml\fR
.
.IP "" 4
.
.nf

<?xml version="1\.0"?>
<!DOCTYPE service_bundle SYSTEM "/usr/share/lib/xml/dtd/service_bundle\.dtd\.1">
<!\-\-
Manifest automatically generated by smfgen\.
\-\->
<service_bundle type="manifest" name="application\-daves\-service" >
    <service name="application/daves\-service" type="service" version="1" >
        <create_default_instance enabled="true" />
        <dependency name="dep1" grouping="require_all" restart_on="error" type="service" >
            <service_fmri value="svc:/milestone/multi\-user:default" />
        </dependency>
        <method_context>
            <method_environment>
                <envvar name=\'PATH\' value=\'/opt/local/bin:/opt/local/sbin:/bin:/usr/bin:/usr/sbin\'/>
            </method_environment>
        </method_context>
        <exec_method type="method" name="start" exec="node /home/dave/daves\-service/server\.js &amp;" timeout_seconds="10" />
        <exec_method type="method" name="stop" exec=":kill" timeout_seconds="30" />
        <template >
            <common_name >
                <loctext xml:lang="C" >Dave\'s Service</loctext>
            </common_name>
        </template>
    </service>
</service_bundle>
.
.fi
.
.IP "" 0
.
.P
This is a fairly standard manifest, generated with smfgen \fIhttps://github\.com/davepacheco/smfgen\fR\. The server code written in Node looks like this:
.
.P
\fBserver\.js\fR
.
.IP "" 4
.
.nf

var http = require(\'http\');

var host = \'localhost\';
var port = 8000;

http\.createServer(onrequest)\.listen(port, host, listening);

function listening() {
  console\.log(\'server started: http://%s:%d/\', host, port);
}

var i = 0;
function onrequest(req, res) {
  console\.log(\'new request\');
  if (++i < 5) res\.end();
}
.
.fi
.
.IP "" 0
.
.P
This webserver has a bug however, on the 5th request and thereafter, all requests will stall, and the server will lock up\. SMF by itself won\'t be able to detect this problem, as the process is up and has a valid \fB/proc\fR structure\. This is where \fBsmfwatchdog\fR can be used to detect and mitigate this issue\.
.
.P
Modify the \fBexec\fR line in the manifest to look like this:
.
.IP "" 4
.
.nf

< <exec_method type="method" name="start" exec="node /home/dave/daves\-service/server\.js &amp;" timeout_seconds="10" />
\-\-\-
> <exec_method type="method" name="start" exec="smfwatchdog &amp; node /home/dave/daves\-service/server\.js &amp;" timeout_seconds="10" />
.
.fi
.
.IP "" 0
.
.P
This instructs the service to start an instance of \fBsmfwatchdog\fR with the node process\.
.
.P
Now, reimport the manifest, and restart the service\. The watchdog will be running as part of the service, which you can verify with \fBsvcs(1M)\fR
.
.IP "" 4
.
.nf

$ svcs \-p daves\-service
STATE          STIME    FMRI
online         10:51:43 svc:/application/daves\-service:default
               10:51:43    32188 smfwatchdog
               10:51:43    32189 node
.
.fi
.
.IP "" 0
.
.P
The watchdog is running, however it has nothing to do, as it has no checks\. We now have to write a health check for the service, to ensure that the webserver is reachable without timing out\.
.
.P
The daemon has created a unique directory for scripts to be placed in \fB/opt/local/share/smf/smfwatchdog/application\-daves\-service:default\fR, you can find that information by looking at the logs of the service\. \fBsmfwatchdog\fR will log to the default log location as found in \fBsvcs \-L <fmri>\fR\.
.
.IP "" 4
.
.nf

$ grep \'plugins directory:\' "$(svcs \-L daves\-service)" | tail \-1
[smfwatchdog@0\.0\.0] [2013\-06\-12T17:51:43\.081Z] plugins directory: /opt/local/share/smf/smfwatchdog/application\-daves\-service:default
.
.fi
.
.IP "" 0
.
.P
Any scripts in this directory will be executed every 60 seconds, and if any of them return with a non\-zero exit code, the service will be restarted, and optionally an email will be sent out alerting of the failed health check and the action taken (including the output generated by the script that failed)\.
.
.P
Since checks are just scripts, we can use any language that we\'d like, so let\'s keep it simple and use bash\. We\'ll create a basic health check to ensure the service is responsive over HTTP\.
.
.IP "" 4
.
.nf

vim /opt/local/share/smf/smfwatchdog/application\-daves\-service\e:default/check\.sh


#!/usr/bin/env bash
CURLE_OPERATION_TIMEDOUT=28
timeout=20 # seconds

curl \-sSk \-m "$timeout" "http://localhost:8000"
if (( $? == $CURLE_OPERATION_TIMEDOUT )); then
        exit 1
else
        exit 0
fi
.
.fi
.
.IP "" 0
.
.P
And ensure the file is executable with:
.
.IP "" 4
.
.nf

chmod +x /opt/local/share/smf/smfwatchdog/application\-daves\-service\e:default/check\.sh
.
.fi
.
.IP "" 0
.
.P
The watchdog daemon will scan the directory every 60 seconds, executing every script it finds, in \fBreaddir(3C)\fR order, and restart the service if any of the scripts exit with a non\-zero exit code\.
.
.P
If curl returns with code 28, that means it has timedout\. In the above script, curl is set to timeout if 20 seconds have elapsed with no response from the server, and then the script itself will return with exit code 1
.
.P
When \fBsmfwatchdog\fR sees that this health check has failed, it will send itself a \fBSIGABRT\fR signal, which will trigger a core dump, and cause the entire service to be restarted by SMF (under most circumstances), see \fIoptions\fR below for different actions to take upon failure\.
.
.P
We can see this happen in the logfile
.
.IP "" 4
.
.nf

tail "$(svcs \-L daves\-service)"
[ Jun 12 17:51:43 Executing start method ("smfwatchdog & node /home/dave/daves\-service/server\.js &")\. ]
[smfwatchdog@0\.0\.0] [2013\-06\-12T17:51:43\.080Z] SMF_FMRI=svc:/application/daves\-service:default
[smfwatchdog@0\.0\.0] [2013\-06\-12T17:51:43\.081Z] plugins directory: /opt/local/share/smf/smfwatchdog/application\-daves\-service:default
[ Jun 12 17:51:43 Method "start" exited with status 0\. ]
server started: http://localhost:8000/
new request
new request
new request
new request
new request
[smfwatchdog@0\.0\.0] [2013\-06\-12T17:57:03\.612Z] check\.sh failed (exit code 1)
[smfwatchdog@0\.0\.0] [2013\-06\-12T17:57:03\.612Z] raising SIGABRT
[ Jun 12 17:57:03 Stopping because process dumped core\. ]
[ Jun 12 17:57:03 Executing stop method (:kill)\. ]
.
.fi
.
.IP "" 0
.
.P
Note that \fBnew request\fR is printed 5 times, as on the 5th time the server will become unresponsive, and the health check will fail\.
.
.P
Optionally, we can set an email address to alert any failures to by setting the environmental variable \fBSMFWATCHDOG_EMAIL\fR in the manifest\.
.
.IP "" 4
.
.nf

<envvar name=\'SMFWATCHDOG_EMAIL\' value=\'dave@daveeddy\.com\'/>
.
.fi
.
.IP "" 0
.
.P
Now, when a health check fails it\'ll fire an informative email containing the output from the health check script that failed, as well as some system information that looks like:
.
.IP "" 4
.
.nf

To: dave@daveeddy\.com
From: noreply@dave\-01\.local
Subject: [smfwatchdog] daves\-service:default failed health check on dave\-01\.local

daves\-service:default failed health check on dave\-01\.local

FMRI: svc:/application/daves\-service:default
Action: raising SIGABRT
Hostname: dave\-01\.local
Time (UTC): 2013\-06\-12T04:38:13
Command: check\.sh
Program: smfwatchdog@0\.0\.0 (compiled Jun 11 2013 21:32:47)

Command Output
curl: (28) Operation timed out after 20000 milliseconds with 0 bytes received
.
.fi
.
.IP "" 0
.
.P
 \fI\fR
.
.SH "Options"
The following options can be passed in as environmental variables, most likely being added to the SMF manifest\.
.
.IP "\(bu" 4
\fBSMFWATCHDOG_DEBUG\fR: (int) If this is non\-zero, \fBsmfwatchdog\fR will produce debug output to the service\'s log file (\fBsvcs \-L <fmri>\fR)
.
.IP "\(bu" 4
\fBSMFWATCHDOG_SLEEP\fR: (int) The time, in seconds, to sleep between running health check scripts, defaults to 60
.
.IP "\(bu" 4
\fBSMFWATCHDOG_ACTION\fR: (int) The action (see below) to take during a health check failure, defaults to 0 (raise \fBSIGABRT\fR)
.
.IP "\(bu" 4
\fBSMFWATCHDOG_IGNORE_CONTRACT\fR: (int) If this is non\-zero, \fBsmfwatchdog\fR will not exit if it is the only process running in its contract
.
.IP "\(bu" 4
\fBSMFWATCHDOG_UID\fR: (int) If this is non\-zero, \fBsetuid(2)\fR will be called with this variable before any plugins are run to drop privileges
.
.IP "\(bu" 4
\fBSMFWATCHDOG_GID\fR: (int) If this is non\-zero, \fBsetgid(2)\fR will be called with this variable before any plugins are run to drop privileges
.
.IP "\(bu" 4
\fBSMFWATCHDOG_COMMAND\fR: A command to execute (parsed by a shell) after a failure case but before any action is taken, defaults to nothing
.
.IP "\(bu" 4
\fBSMFWATCHDOG_EMAIL\fR: If set, this variable will be used as an email address to send alerts to when a service has failed a health check
.
.IP "\(bu" 4
\fBSMFWATCHDOG_EMAIL_FROM\fR: This is the address from which the above email will be sent, defaults to \fBnoreply@<hostname>\fR
.
.IP "\(bu" 4
\fBSMFWATCHDOG_MAIL_PROG\fR: The mail program to use to send email on the system, it must accept binary email data over stdin, defaults to \fBmailx \-t\fR
.
.IP "\(bu" 4
\fBSMFWATCHDOG_DISABLED\fR: If set, \fBsmfwatchdog\fR will exit cleanly upon starting
.
.IP "\(bu" 4
\fBSMF_FMRI\fR: This shouldn\'t be manually set, it will be set automatically by SMF, and is used to tell the watchdog which service to monitor
.
.IP "" 0
.
.SS "Actions"
You can set the action to be taken during a health check failure by setting \fBSMFWATCHDOG_ACTION\fR to a valid integer listed below\.
.
.IP "" 4
.
.nf

#define ACT_RAISE_SIGABRT  0   /* kill ourself with SIGABRT */
#define ACT_RESTART_SVC    1   /* restart our own service (requires priv) */
#define ACT_EXIT           2   /* exit with a failure error code */
#define ACT_NOTHING        3   /* do nothing */
.
.fi
.
.IP "" 0
.
.P
The default action is to raise a \fBSIGABRT\fR signal to trigger a core dump\. This will cause SMF to restart the entire service, without \fBsmfwatchdog\fR requiring escalated privileges\.
.
.P
Setting this variable to 1 will cause the command \fBsvcadm restart <fmri>\fR to be triggered after a failed health check\. Note that this will only work if the effective UID of the \fBsmfwatchdog\fR daemon has privileges to carry out that command\.
.
.P
Setting this variable to 2 will cause \fBsmfwatchdog\fR to exit with a failure return code after a failed check\.
.
.P
Setting this variable to 3 will cause \fBsmfwatchdog\fR to not take any action except to log and optionally send an email in the event of a failed health check, making it a good option for testing\.
.
.P
 \fI\fR
.
.SH "Usage"
If you run \fBsmfwatchdog\fR interactively (without \fBSMF_FMRI\fR set) you are greeted with
.
.IP "" 4
.
.nf

$ smfwatchdog
smfwatchdog is not meant to be run interatively
.
.fi
.
.IP "" 0
.
.P
as the daemon would have nothing to check\.
.
.P
If \fBsmfwatchdog\fR is run with any number of arguments, the version string is printed and the process exits cleanly\.
.
.IP "" 4
.
.nf

$ smfwatchdog \-v
smfwatchdog@0\.0\.2 (compiled Jun 13 2013 09:09:46)
.
.fi
.
.IP "" 0
.
.P
The proper way to run \fBsmfwatchdog\fR is to add it to the exec line of an SMF manifest, and let SMF start and stop the daemon\.
.
.P
 \fI\fR
.
.SH "Notes"
The watchdog isn\'t a solution to a problem; it doesn\'t fix bugs\. The health checks are meant to minimize the impact of bugs that exist that currently have not been fixed in a service\.
.
.IP "\(bu" 4
If \fBsmfwatchdog\fR is the only process running in its contract, it will exit with an error code to let SMF run its course
.
.IP "" 0
.
.SS "Concerns"
.
.IP "\(bu" 4
This software hasn\'t been fully tested, and may contain bugs
.
.IP "\(bu" 4
Scripts are executed with \fBpopen(3C)\fR, and as such, have their names parsed by the shell\. Ensure that you don\'t name your scripts \fB$(rm \-rf /)\fR or something
.
.IP "\(bu" 4
Don\'t set \fBSMFWATCHDOG_UID\fR or \fBSMFWATCHDOG_GID\fR if the manifest itself takes care of dropping privileges
.
.IP "" 0
.
.P
 \fI\fR
.
.SH "Todo"
.
.IP "\(bu" 4
\fBpclose\fR raises \fBSIGABRT\fR when executing a broken symlink\.
.
.IP "" 0
.
.P
 \fI\fR
.
.SH "License"
MIT License


smfwatchdog.c
/**
 * SMF Watch Dog
 *
 * A health checking daemon to be used with an
 * SMF service.
 *
 * Author: Dave Eddy <dave@daveeddy.com>
 * Date: 5/19/2013
 * License: MIT
 */

#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <libcontract.h>
#include <libgen.h>
#include <netdb.h>
#include <procfs.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

#define PROGNAME           "smfwatchdog"
#define VERSION            "0.0.7"

#define DATEFMT            "%Y-%m-%dT%H:%M:%S"
#define DEFAULT_MAIL_PROG  "mailx -t"
#define MAX_FMRI_LEN       scf_limit(SCF_LIMIT_MAX_FMRI_LENGTH)
#define WATCHDOG_DIR       "/opt/local/share/smf/smfwatchdog"

/* Action to take on failure, set by SMFWATCHDOG_ACTION */
#define ACT_RAISE_SIGABRT  0   /* kill ourself with SIGABRT */
#define ACT_RESTART_SVC    1   /* restart our own service (requires priv) */
#define ACT_EXIT           2   /* exit with a failure error code */
#define ACT_NOTHING        3   /* do nothing */

/* LOG() if SMFWATCHDOG_DEBUG is set */
#define DEBUG(...) \
	do { \
		if (options.debug) LOG(__VA_ARGS__); \
	} while (0)

/* Environmental Options */
struct {
	int debug;       /* SMFWATCHDOG_DEBUG
	                    whether debug output is enabled */
	int sleep;       /* SMFWATCHDOG_SLEEP
	                    time to sleep between looping all the checks */
	int action;      /* SMFWATCHDOG_ACTION
			    the action to take when there is a failure */
	int no_contract; /* SMFWATCHDOG_IGNORE_CONTRACT
			    if set, the contract will not be looked up */
	int uid;         /* SMFWATCHDOG_UID
			    uid used to drop privileges before looping */
	int gid;         /* SMFWATCHDOG_GID
			    gid used to drop privileges before looping */
	char *command;   /* SMFWATCHDOG_COMMAND
			    a command to run before any action is taken */
	char *mail_to;   /* SMFWATCHDOG_EMAIL
			    an email address to alert when a health check
			    fails */
	char *mail_from; /* SMFWATCHDOG_EMAIL_FROM
			    who the email originates from, defaults to
			    noreply@<hostname> */
	char *mail_prog; /* SMFWATCHDOG_MAIL_PROG
			    a program that accepts mail data over stdin
			    to send an email, defaults to "mailx -t" */
} options;

/* Function Prototypes */
void LOG(const char *fmt, ...);
void loadenvironment();
int done(int code);
int execute(const char *cmd, char **output);
int process();
int strreplace(char *s, char f, char r);
int sendmail(const char *check, const char *body);
int contract_id_by_pid(pid_t pid);
int num_pids_in_contract(int contractid);

int main(int argc, char **argv) {
	setbuf(stdout, NULL);

	/* print the version string if run with any arguments */
	if (argc > 1) {
		printf("%s@%s (compiled %s %s)\n",
		    PROGNAME, VERSION, __DATE__, __TIME__);
		return 0;
	}

	/* get the SMF FMRI */
	char *FMRI = getenv("SMF_FMRI");
	if (FMRI == NULL || FMRI[0] == '\0') {
		printf("%s is not meant to be run interatively\n\n",
		    PROGNAME);
		return 1;
	}

	/* check if we are disabled */
	char *disabled = getenv("SMFWATCHDOG_DISABLED");
	if (disabled != NULL && disabled[0] != '\0') {
		LOG("SMFWATCHDOG_DISABLED is set\n");
		return done(0);
	}

	LOG("SMF_FMRI=%s\n", FMRI);

	/* create the SMF watchdog dir if it doesn't exist */
	mkdir(WATCHDOG_DIR, 0755);

	int FMRI_LEN = strlen(FMRI);
	char _name[FMRI_LEN + 1];
	char *name = _name;

	/* copy the FMRI and remove the svc:/ prefix */
	strcpy(name, FMRI);
	if (strncmp(name, "svc:/", 5) == 0) {
		name += 5;
	}
	/* replace / with - */
	strreplace(name, '/', '-');

	/* plugins directory */
	/* <base> + '/' + <name> + '\0' */
	char dir[strlen(WATCHDOG_DIR) + 1 + strlen(name) + 1];
	sprintf(dir, "%s/%s", WATCHDOG_DIR, name);

	LOG("plugins directory: %s\n", dir);
	mkdir(dir, 0755);

	/* test dir existence by moving into it */
	if (chdir(dir) != 0) {
		LOG("chdir(%s): %s\n", dir, strerror(errno));
		return done(2);
	}

	/* load env options */
	loadenvironment();

	/* create the restart cmd */
	char restartcmd[15 + FMRI_LEN + 1];
	sprintf(restartcmd, "svcadm restart %s", FMRI);

	/* drop privileges */
	if (options.gid && setgid(options.gid) < 0) {
		LOG("setgid to %d failed: %s\n",
		    options.gid, strerror(errno));
		return done(3);
	}
	if (options.uid && setuid(options.uid) < 0) {
		LOG("setuid to %d failed: %s\n",
		    options.uid, strerror(errno));
		return done(4);
	}

	/* get the contract id of this process */
	int contractid = contract_id_by_pid(getpid());
	if (contractid < 0) {
		LOG("failed to get contract id: %s\n", strerror(errno));
		if (!options.no_contract)
			return done(5);
	}
	LOG("contract id: %d\n", contractid);

	/* start the loop */
	int ret = 0;
	int loop = 1;
	int numpids = 0;
	while (loop) {
		DEBUG("sleeping for %d seconds\n", options.sleep);
		sleep(options.sleep);

		DEBUG("waking up from sleep\n");

		/* check to make sure we aren't the only process in the contract */
		if (!options.no_contract) {
			numpids = num_pids_in_contract(contractid);
			DEBUG("pids found in this contract: %d\n", numpids);
			if (numpids == 1) {
				LOG("last process running in this contract, exiting\n");
				return done(6);
			}
		}

		/* loop the directories */
		ret = process();
		if (ret == 0) continue;

		/* If we are here, something failed */
		if (options.command != NULL) {
			LOG("executing: %s\n", options.command);
			system(options.command);
		}
		switch (options.action) {
			default:
				LOG("unknown action\n");
			case ACT_RAISE_SIGABRT:
				LOG("raising SIGABRT\n");
				raise(SIGABRT);
				break;
			case ACT_RESTART_SVC:
				/* "svcadm restart " + FMRI + "\0" */
				LOG("executing: %s\n", restartcmd);
				system(restartcmd);
				break;
			case ACT_EXIT:
				loop = 0;
				break;
			case ACT_NOTHING:
				break;
		}
	}

	/* we have broken from the loop, exit */
	return done(ret);
}

/**
 * print "exiting." and return the code you give it
 */
int done(int code) {
	LOG("exiting.\n");
	return code;
}

/**
 * load environmental variables into the options struct
 */
void loadenvironment() {
	char *p = NULL;

	p = getenv("SMFWATCHDOG_DEBUG");
	options.debug = (p == NULL) ? 0 : atoi(p);
	DEBUG("option: {SMFWATCHDOG_DEBUG} debug output %s\n",
	    options.debug ? "enabled" : "disabled");

	p = getenv("SMFWATCHDOG_SLEEP");
	if (p != NULL) options.sleep = atoi(p);
	if (!options.sleep) options.sleep = 60;
	DEBUG("option: {SMFWATCHDOG_SLEEP} sleep %d seconds\n",
	    options.sleep);

	p = getenv("SMFWATCHDOG_ACTION");
	if (p != NULL) options.action = atoi(p);
	DEBUG("option: {SMFWATCHDOG_ACTION} on failure action %d\n",
	    options.action);

	p = getenv("SMFWATCHDOG_IGNORE_CONTRACT");
	if (p != NULL) options.no_contract = atoi(p);
	DEBUG("option: {SMFWATCHDOG_IGNORE_CONTRACT} ignore contract %d\n",
	    options.no_contract);

	p = getenv("SMFWATCHDOG_UID");
	if (p != NULL) options.uid = atoi(p);
	DEBUG("option: {SMFWATCHDOG_UID} uid for dropped privileges %d\n",
	    options.uid);

	p = getenv("SMFWATCHDOG_GID");
	if (p != NULL) options.gid = atoi(p);
	DEBUG("option: {SMFWATCHDOG_GID} gid for dropped privileges %d\n",
	    options.gid);

	options.command = getenv("SMFWATCHDOG_COMMAND");
	DEBUG("option: {SMFWATCHDOG_COMMAND} command to run \"%s\"\n",
	    options.command);

	options.mail_to = getenv("SMFWATCHDOG_EMAIL");
	DEBUG("option: {SMFWATCHDOG_EMAIL} email to \"%s\"\n",
	    options.mail_to);

	options.mail_from = getenv("SMFWATCHDOG_EMAIL_FROM");
	DEBUG("option: {SMFWATCHDOG_EMAIL_FROM} email from \"%s\"\n",
	    options.mail_from);

	p = getenv("SMFWATCHDOG_MAIL_PROG");
	options.mail_prog = (p != NULL && p[0] != '\0')
	    ? p : DEFAULT_MAIL_PROG;
	DEBUG("option: {SMFWATCHDOG_MAIL_PROG} mail prog \"%s\"\n",
	    options.mail_prog);
}

/**
 * loop over all files in the current directory and execute them
 */
int process() {
	struct dirent *dp; /* dir pointer */
	DIR *d = opendir(".");

	if (d == NULL) {
		LOG("opendir(): %s\n", strerror(errno));
		return 1;
	}

	/* loop over the dirent */
	int i = 0;
	int ret = 0;
	while ((dp = readdir(d)) != NULL) {
		/* skip hidden files */
		if (dp->d_name[0] == '.') continue;

		DEBUG("executing %s\n", dp->d_name);

		/* run the plugin */
		char *output = NULL;
		ret = execute(dp->d_name, &output);
		if (options.debug && output != NULL)
			printf("%s", output);

		/* something went wrong, ignore it */
		if (ret == -1) {
			LOG("error executing %s, moving on\n", dp->d_name);
			if (output != NULL) free(output);
			continue;
		}
		ret = WEXITSTATUS(ret);

		switch (ret) {
			case 0: /* success */
				DEBUG("%s executed succesfully\n",
				    dp->d_name);
				break;
			case 126: /* permission denied, consider this a success */
				break;
			default: /* health check failed */
				LOG("%s failed (exit code %d)\n",
				    dp->d_name, ret);

				if (options.mail_to != NULL) {
					LOG("sending email to %s\n",
					    options.mail_to);
					sendmail(dp->d_name, output);
				}
				if (output != NULL) free(output);
				return ret;
		}
		if (output != NULL) free(output);
		i++;
	}
	closedir(d);

	DEBUG("%d scripts executed\n", i);

	return 0;
}

/**
 * execute the given script in the current directory
 */
int execute(const char *cmd, char **output) {
	/* "exec ./" + dp->d_name + " 2>&1\0" */
	char prog[7 + strlen(cmd) + 6];
	sprintf(prog, "exec ./%s 2>&1", cmd);

	FILE *fp = popen(prog, "r");
	if (fp == NULL) {
		LOG("popen: %s\n", strerror(errno));
		return -1;
	}

	char buf[BUFSIZ]; /* output buffer */
	int bytes = 0; /* bytes read */
	int outputsize = 1; /* output buffer size (+1 for nul byte) */
	char *out = malloc(sizeof(*out) * outputsize); /* all cmd output */
	if (out == NULL) {
		LOG("malloc: %s\n", strerror(errno));
		return -1;
	}
	out[0] = '\0';

	/* save all of the cmd output */
	while ((bytes = fread(buf, sizeof(*buf), sizeof(buf) / sizeof(*buf), fp)) > 0) {
		outputsize += bytes;
		char *tmp = realloc(out, outputsize);
		if (tmp == NULL) {
			LOG("realloc: %s\n", strerror(errno));
			free(out);
			out = NULL;
			return -1;
		}
		out = tmp;
		strncat(out, buf, bytes);
		out[outputsize - 1] = '\0';
	}
	if (!feof(fp)) {
		LOG("couldn't read until EOF: %s\n", strerror(errno));
		return -1;
	} else if (ferror(fp)) {
		LOG("ferror: %s\n", strerror(errno));
		return -1;
	}

	*output = out;
	return pclose(fp);
}

/**
 * print a timestamped log line
 *
 * usage is the same as printf()
 *
 * returns nothing
 */
void LOG(const char *fmt, ...) {
	char date[20];
	struct timeval tv;
	va_list args;

	/* print the progname, version, and timestamp */
	gettimeofday(&tv, NULL);
	strftime(date, sizeof(date) / sizeof(date[0]), DATEFMT, gmtime(&tv.tv_sec));
	printf("[%s@%s] [%s.%03ldZ] ", PROGNAME, VERSION, date, tv.tv_usec / 1000);

	/* printf like normal */
	va_start(args, fmt);
	vprintf(fmt, args);
	va_end(args);
}

/**
 * replace characters in a string
 *
 * replace all occurences of char f with char r in char * s
 *
 * returns     number of replacements done
 */
int strreplace(char *s, char f, char r) {
	int i = 0;
	char *p = s;
	while (*p != '\0') {
		if (*p == f) {
			*p = r;
			i++;
		}
		p++;
	}
	return i;
}

/**
 * send an email message
 *
 * const char *check   the name of the check that failed
 * const char *body    the email body, rendered as HTML
 *
 * return         int pclose()
 */
int sendmail(const char *check, const char *body) {
	char *FMRI = getenv("SMF_FMRI");
	char *base = basename(FMRI);

	/* open the mail program for writing binary data */
	FILE *email = popen(options.mail_prog, "wb");
	if (email == NULL) {
		LOG("failed to email \"%s\" using program \"%s\": %s\n",
		    options.mail_to, options.mail_prog, strerror(errno));
		return -1;
	}

	/* get the hostname */
	char hostname[MAXHOSTNAMELEN + 1];
	hostname[MAXHOSTNAMELEN] = '\0';
	if (gethostname(hostname, MAXHOSTNAMELEN) == -1) {
		LOG("gethostname(): %s\n", strerror(errno));
		strcpy(hostname, "unknown");
	}

	/* noreply@ + '\0' + <hostname> + '\0' */
	char mail[8 + MAXHOSTNAMELEN + 1];
	mail[8 + MAXHOSTNAMELEN] = '\0';
	if (options.mail_from == NULL) {
		/* create a mail_from address */
		sprintf(mail, "noreply@%s", hostname);
	}

	char action[256];
	/* find out what action will be taken */
	switch (options.action) {
		default:
		case ACT_RAISE_SIGABRT:
			snprintf(action, sizeof(action) / sizeof(action[0]),
			    "raising SIGABRT");
			break;
		case ACT_RESTART_SVC:
			snprintf(action, sizeof(action) / sizeof(action[0]),
			    "restarting service with <code>svcadm restart %s</code>",
			    FMRI);
			break;
		case ACT_EXIT:
			snprintf(action, sizeof(action) / sizeof(action[0]),
			    "process exiting");
			break;
		case ACT_NOTHING:
			snprintf(action, sizeof(action) / sizeof(action[0]),
			    "no action taken");
			break;
	}

	/* get the timestamp */
	char date[20];
	struct timeval tv;
	gettimeofday(&tv, NULL);
	strftime(date, sizeof(date) / sizeof(date[0]), DATEFMT, gmtime(&tv.tv_sec));

	fprintf(email, "To: %s\n", options.mail_to);
	fprintf(email, "From: %s\n",
	    options.mail_from == NULL ?  mail : options.mail_from);
	fprintf(email, "Subject: [%s] %s failed health check on %s\n",
	    PROGNAME, base, hostname);
	fprintf(email, "Content-Type: text/html\n");
	fprintf(email, "\n");
	fprintf(email, "<code>%s</code> failed health check on <code>%s</code><br><br>\n\n",
	    base, hostname);
	fprintf(email, "<b>FMRI:</b> <code>%s</code><br>\n", FMRI);
	fprintf(email, "<b>Action:</b> <code>%s</code><br>\n", action);
	fprintf(email, "<b>Hostname:</b> <code>%s</code><br>\n", hostname);
	fprintf(email, "<b>Time (UTC):</b> <code>%s</code><br>\n", date);
	fprintf(email, "<b>Command:</b> <code>%s</code><br>\n", check);
	fprintf(email, "<b>Program:</b> <code>%s@%s (compiled %s %s)</code><br><br>\n\n",
	    PROGNAME, VERSION, __DATE__, __TIME__);

	fprintf(email, "<b>Command Output</b>\n");
	fprintf(email, "<pre>%s</pre>\n", body);

	return pclose(email);
}

/**
 * get the contract id of a given pid
 *
 * @param pid {pid_t} pid to check
 *
 * @returns the contract id, or a negative number
 * with errno set on failure
 */
int contract_id_by_pid(pid_t pid) {
	if (!pid) pid = getpid();

	struct psinfo info; /* psinfo struct for the process */
	char psinfo_file[256]; /* /proc/<pid>/psinfo */
	int fd; /* reusable fd */

	snprintf(psinfo_file,
		sizeof(psinfo_file) / sizeof(*psinfo_file),
		"/proc/%d/psinfo",
		(int)pid);

	/* read psinfo and load the struct*/
	fd = open(psinfo_file, O_RDONLY);
	if (fd < 0)
		return -1;
	if (read(fd, &info, sizeof(info)) != sizeof(info)) {
		close(fd);
		return -2;
	}
	close(fd);

	/* the processes contract id */
	return (int)info.pr_contract;
}

/**
 * Checks how many pid's exist in a given contract
 *
 * @param contractid {int} the contract ID to check
 *
 * @returns the number of pids in a contract, or a negative number
 * with errno set on failure
 */
int num_pids_in_contract(int contractid) {
	char contract_file[256]; /* /system/contract/all/<ctid>/status */
	int fd; /* reusable fd */

	/* a contract stat handle */
	ct_stathdl_t stathdl;

	snprintf(contract_file,
		sizeof(contract_file) / sizeof(*contract_file),
		"/system/contract/all/%d/status",
		contractid);
	fd = open(contract_file, O_RDONLY | O_LARGEFILE);
	if (fd < 0)
		return -3;
	if (ct_status_read(fd, CTD_ALL, &stathdl) != 0) {
		close(fd);
		return -4;
	}
	close(fd);

	pid_t *members;
	uint_t numpids;
	int err = 0;
	if ((err = ct_pr_status_get_members(stathdl, &members, &numpids)))
		return -5;
	ct_status_free(stathdl);

	return numpids;
}
